libyul/optimiser/FunctionDefinitionCollector.h

```diff
@@ -0,0 +1,44 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * AST walker that finds all function definitions and stores them into a map indexed by the function names.
+ */
+#pragma once
+
+#include <libyul/optimiser/ASTWalker.h>
+
+#include <map>
+
+namespace solidity::yul
+{
+
+/**
+ * AST walker that finds all function definitions and stores them into a map indexed by the function names.
+ *
+ * Prerequisite: Disambiguator
+ */
+class FunctionDefinitionCollector: ASTWalker
```

Comment:
Does this need to be an AST walker? It could also be a filter on a vector of statements.

Reply:
(assuming the function hoister)

Reply:
Yeah, I guess that would be enough... but is that worth the danger of running it on non-function-hoisted code? If I want to validate that the input is hoisted, I can just as well visit the entire AST anyways...
But I'd be fine with changing it (respectively removing the analysis step and just walking the outermost vector instead).

Reply:
Isn't the hoister a requirement for everything? But yeah, I'm undecided.